### ISSUE TITLE
Feature: File upload?

### ISSUE THREAD

**Issue by stubailo** [apollographql/apollo-fetch#6]
I think `apollo-fetch` has an opportunity to be the only simple GraphQL client library out there with support for file uploads. Looks like `apollo-upload-client` is quite simple implementation-wise, so it would be fine to build that stuff right in!

@jaydenseric @evanshauser @jbaxleyiii let's talk here, I think this would be a big win not only for Apollo Client having file upload capability, but also for `apollo-fetch` as a standalone utility!

**Comment by stubailo** [comment]
(Note: the final decision on the best path forward here is up to @evanshauser and @jbaxleyiii, I'm just facilitating!)

**Comment by jbaxleyiii** [comment]
@jaydenseric I 💯 think we should include your work in `apollo-fetch`. Let me know how I can help you out!

**Comment by jaydenseric** [comment]
Cool, I'll get familiar with the codebase and the middleware/afterware system.

**Comment by stubailo** [comment]
For reference: https://github.com/jaydenseric/apollo-upload-client

**Comment by jaydenseric** [comment]
Is query batching a consideration here?

**Comment by evanshauser** [comment]
@jaydenseric Currently, batching could be done in a custom fetch function(which merges the http specific headers) or in another stand-alone library(middleware/afterware have access to the batch). Does either option make file upload easier for you now or for a later transition? 

For now though, servicing a single upload works great! I envision the manipulations would be made before the middleware, so that they have the ability to modify the full request. Or file upload could be a middleware that is exported with the apollo-fetch. Let me know if that fits with your thoughts or you like something better.

CC @jbaxleyiii

**Comment by jaydenseric** [comment]
We have a choice: Run middleware before, or after the upload process transforms the request body to `FormData` for a multipart upload.

If middleware runs before, it will be able to access the request as an object whether uploads are present or not. Middleware would not have the last say on what gets sent.

If it runs after, authors will need to account for the request being a `FormData` instance sometimes. Middleware would have the last say.

I think it's best to run middleware before uploads are processed.

As the codebase is currently, it would not be possible to implement uploads via middleware. The request body is [forcibly converted to a JSON string](https://github.com/apollographql/apollo-fetch/blob/master/src/apollo-fetch.ts#L73). We could change that to first check if the request is an instance of `FormData` or just not use middleware to transform the request.

To get client part done here, the easiest approach is to make it compatible with `apollo-upload-server` so we can see it working. The way the request is transformed for transport between client and server is really tricky, particularly factoring in batching. I will try to document the transformations with examples here so we can be sure everyone is happy because it will become a bit of a defacto standard, like query batching with Apollo server is.

**Comment by jaydenseric** [comment]
Noting here the current spec for request transforms between `apollo-upload-client` and `apollo-upload-server`.

This is just a JSON representation of what's going on each step; I have tidied the order of some of the properties, omitted others that we discard anyway (File fields in FormData have a lot going on), and used JSON to represent instances.

You can use [jaydenseric/apollo-upload-examples](https://github.com/jaydenseric/apollo-upload-examples) to log this behavior. I am working on a new [`apollo-fetch` branch](https://github.com/jaydenseric/apollo-upload-examples/tree/apollo-fetch).

## No query batching

### Before `apollo-upload-server`

```json
{
  "operations": "{\"query\":\"mutation multipleUpload($files: [Upload!]!) {\\n  multipleUpload(files: $files) {\\n    id\\n    name\\n    type\\n    size\\n    path\\n    __typename\\n  }\\n}\\n\",\"variables\":{\"files\":[null,null]},\"operationName\":\"multipleUpload\"}",
  "variables.files.0": {
    "name": "test-file-1.png",
    "type": "image/png",
    "size": 44631,
    "path": "/tmp/apollo-upload-examples/upload_ca4fb54e9e8edf655b90764c97c5e28d"
  },
  "variables.files.1": {
    "name": "test-file-2.pdf",
    "type": "application/pdf",
    "size": 13472,
    "path": "/tmp/apollo-upload-examples/upload_6cece2a2285305d340a1d174af544ec6"
  }
}
```

### After `apollo-upload-server`

```json
{
  "operationName": "multipleUpload",
  "query": "mutation multipleUpload($files: [Upload!]!) {\n  multipleUpload(files: $files) {\n    id\n    name\n    type\n    size\n    path\n    __typename\n  }\n}\n",
  "variables": {
    "files": [
      {
        "name": "test-file-1.png",
        "type": "image/png",
        "size": 44631,
        "path": "/tmp/apollo-upload-examples/upload_ca4fb54e9e8edf655b90764c97c5e28d"
      },
      {
        "name": "test-file-2.pdf",
        "type": "application/pdf",
        "size": 13472,
        "path": "/tmp/apollo-upload-examples/upload_6cece2a2285305d340a1d174af544ec6"
      }
    ]
  }
}
```

## With query batching

### Before `apollo-upload-server`

```json
{
  "operations": "[{\"query\":\"mutation multipleUpload($files: [Upload!]!) {\\n  multipleUpload(files: $files) {\\n    id\\n    name\\n    type\\n    size\\n    path\\n    __typename\\n  }\\n}\\n\",\"variables\":{\"files\":[null,null]},\"operationName\":\"multipleUpload\"}]",
  "0.variables.files.0": {
    "name": "test-file-1.png",
    "type": "image/png",
    "size": 44631,
    "path": "/tmp/apollo-upload-examples/upload_c6fa87e18003c9a5013d6ae33779aa6d"
  },
  "0.variables.files.1": {
    "name": "test-file-2.pdf",
    "type": "application/pdf",
    "size": 13472,
    "path": "/tmp/apollo-upload-examples/upload_dfe8315d7b78b9011998ec07a43217e3"
  }
}
```

### After `apollo-upload-server`

```json
[
  {
    "operationName": "multipleUpload",
    "query": "mutation multipleUpload($files: [Upload!]!) {\n  multipleUpload(files: $files) {\n    id\n    name\n    type\n    size\n    path\n    __typename\n  }\n}\n",
    "variables": {
      "files": [
        {
          "name": "test-file-1.png",
          "type": "image/png",
          "size": 44631,
          "path": "/tmp/apollo-upload-examples/upload_c6fa87e18003c9a5013d6ae33779aa6d"
        },
        {
          "name": "test-file-2.pdf",
          "type": "application/pdf",
          "size": 13472,
          "path": "/tmp/apollo-upload-examples/upload_dfe8315d7b78b9011998ec07a43217e3"
        }
      ]
    }
  }
]
```

**Comment by jaydenseric** [comment]
@stubailo @evanshauser @jbaxleyiii I just did a lot of edits to my above comment regarding a spec for multipart post requests to `graphql-server`, that borrows from current behavior of `apollo-upload-client` and `apollo-upload-server`. Please let me know your thoughts so I can get stuck into implementing – I have cleared my calendar for the next few days to work on it 🙂

**Comment by stubailo** [comment]
It's quite late here, so we won't be able to say anything useful until tomorrow. I agree this might not make sense to implement as a middleware.

As one more question, would it make possible to expand the middleware functionality to enable file upload to be implemented in middleware form? Somehow make it possible for the request options set by the middleware contain multipart form data?

If we think there might be multiple ways of encoding and sending files, that might be an advantage. On the other hand, it might also be useful to "enforce" a standard way of sending files so that everyone can be using the same approach.

**Comment by jaydenseric** [comment]
There are a few moving parts here, but the more I think about it, the best thing to start with is to standardize a structure for multipart POST requests between a client and `graphql-server`. Essentially the structure described under the "before apollo-upload-server" headings in my [earlier comment](https://github.com/apollographql/apollo-fetch/issues/6#issuecomment-314039458).

Once that is agreed upon, client and server implementations can be decided separately and worked on immediately.

The details of how a client generates the multipart request (using middleware, etc.) can be worked out. The details of what a server does after receiving the multipart form (store files to a temp dir like `apollo-upload-server`, or try to provide them in resolver arguments as streams, etc.) can be worked out later or even be pluggable.

In the short term it would be a win for users if they only had to use `apollo-upload-client`, and `graphql-server` would just work. Or the other way around, use `apollo-upload-server` and `apollo-fetch` would just work.

**Comment by evanshauser** [comment]
I really like this idea that we can do batching, since `query` is just a string! We should also think through standardizing how the variables are associated with their specific query.

In terms of standardizing the single multipart POST request format, I hesitate to use the structure of "before apollo-upload-server" over the nesting that "after apollo-upload server" exhibits(minus the temporary paths). In my mind, this nesting feels more natural with apollo-fetch's current interface:

```
{
  "operationName": "multipleUpload",
  "query": "mutation multipleUpload($files: [Upload!]!) {\n  multipleUpload(files: $files) {\n    id\n    name\n    type\n    size\n    path\n    __typename\n  }\n}\n",
  "variables": {
    "files": [
      {
        "name": "test-file-1.png",
        "type": "image/png",
        "size": 44631,
      },
    ],
  }
}
```

I'm not convinced it is better, since it means we wouldn't be immediately compatible with `apollo-upload-server` and `apollo-upload-client` and we would either need to reserve some keywords in the query string or provide a fourth argument sent to the server, such as a context that indicates a file upload:

```
{
  "operationName": "multipleUpload",
  "query": "mutation multipleUpload($files: [Upload!]!) {\n  multipleUpload(files: $files) {\n    id\n    name\n    type\n    size\n    path\n    __typename\n  }\n}\n",
  "variables": {
    "files": [
      {
        "name": "test-file-1.png",
        "type": "image/png",
        "size": 44631,
      },
    ],
  }
  "context": {
    upload: true,
  }
}
```

@jaydenseric what do you think about the standard/does it provide the information the server needs? Would it be reasonable for the upload client and server to follow a similar pattern?

Also curious what @jbaxleyiii thinks

CC @stubailo

**Comment by jaydenseric** [comment]
@evanshauser multipart forms are limited in that the actual files being uploaded can only be top level fields. They have to be extracted from the operation for transport then put back in the right place at the other end. Keep in mind that the file does not actually look like the nice JSON representation when a field in a multipart form, it is a big blob with all the binary data.

There is a lot to explain about how to efficiently process the multipart form stream on the server in relation to resolver arguments, but that can be another conversation.

**Comment by jaydenseric** [comment]
When batching the files from each batch together occupy top level multipart form fields. That is why in the example the field name `0.variables.files.0` has a leading zero – the re-assembly path begins with the batch index.

**Comment by evanshauser** [comment]
Thank you for the pointer! I'll take a look at it in the morning and I have to think about the impact of adding fields to the operation passed to fetch, so I should get back to you in a couple hours. 

Our future plan for apollo-fetch in addition to being a lightweight fetcher is to underlie an http link, which follows the [specification](https://github.com/apollographql/apollo-link/blob/cdc376930aee6e119e4f840527d9d5f045097410/SPEC.md) of apollo-link. This prompts the thought to make a file upload into a link that ends a chain, similar to httpLink. I think it comes down to where we want to put the file upload complexity. Your thoughts on the best place, inside a link or apollo-fetch, would be really informative and most likely guide the decision.

The usage would look like this:

```js
const networkInterface = Links.split(
  isFileUpload,
  new FileUploadLink({uri}),
  new HttpLink({uri}),
);
```

**Comment by jaydenseric** [comment]
I expect to have a PR for `apollo-fetch` ready in the next day or so. I am currently breaking out some `apollo-upload-client` logic that would be useful here and in other clients into a seperate utility npm package.

**Comment by jaydenseric** [comment]
I have published [`extract-files`](https://github.com/jaydenseric/extract-files) and released [`apollo-upload-client` v5.1.0](https://github.com/jaydenseric/apollo-upload-client/releases/tag/v5.1.0).

The logic (and `ReactNativeFile` class) to do with extracting files from the variables object has been pulled out from `apollo-upload-client` into `extract-files` and is now tested.

It should be pretty easy now to [implement it](https://github.com/jaydenseric/apollo-upload-client/blob/master/src/network-interface.js#L9) here. I never use Typescript though so that will really slow me down. Lets see.

**Comment by savovs** [comment]
Hey guys, I'm getting an error:
```
TypeError
message: "Object(...) is not a function"
```

It happens when I do this:
```ts
import { connect } from 'react-redux';
import { createApolloFetchUpload } from 'apollo-fetch-upload';

import Venues, { Props } from '../components/Venues';

client = createApolloFetchUpload({ uri: `...` });

type Venue = {
  id: number;
  name: string;
  email: string;
  address: string;
  description?: string;
  picture?: File;
};

function addVenue(venue: Venue) {
	const query = `
		mutation addVenue($name: String!, $email: String!, $address: String!, $description: String, $picture: File) {
			addVenue(name: $name, email: $email, address: $address, description: $description, picture: $picture) {
				id
			}
		}
	`;

	return client({ query, variables: { ...venue } });
}

export default connect<Props>(
	null, 
	() => ({ addVenue }),
)(Venues);
```

Here's where it breaks:
```js
function constructUploadOptions(requestOrRequests, options) {
    var files = Object(__WEBPACK_IMPORTED_MODULE_1_extract_files__["extractFiles"])(requestOrRequests);
    if (files.length) {
        if (typeof FormData === 'undefined') {
            throw new Error('Environment must support FormData to upload files.');
        }
        options.method = 'POST';
        options.body = new FormData();
        options.body.append('operations', JSON.stringify(requestOrRequests));
        files.forEach(function (_a) {
            var path = _a.path, file = _a.file;
            return options.body.append(path, file);
        });
        return options;
    }
    return Object(__WEBPACK_IMPORTED_MODULE_0_apollo_fetch__["a" /* constructDefaultOptions */])(requestOrRequests, options);
}
```
